### PR TITLE
[AzureStorage] Add optional JsonConverter for OrchestrationInstance

### LINKED ISSUE #491: DTFx.AzureStorage not honoring DataContract on OrchestrationInstance during serializing
`OrchestrationInstance` is marked with `DataContract` and has an `ExtensionData` object on it. This is to allow round-trip serialization when the types do not exactly match on each end, without losing any data. Unfortunately, AzureStorage is using custom Newtonsoft Json serialization with `TypeNameHandling.Object` and an `ISerializationBinder` which drops assembly name. The result is that AzureStorage implementation cannot round-trip any OrchestrationInstance serialization that has a change in type name. This means that any refactor that changes OrchestrationInstance full name will break ongoing orchestrations, and it outright prevents any use of custom OrchestrationInstance implementations (which is possible with other providers, like ServiceBus).

Luckily Newtonsoft.Json does honor `DataContract` attributes already, the fix should be as simple as adding `[JsonProperty(TypeNameHandling = TypeNameHandling.None)]` to `MessageData.Sender`.

### PR DESCRIPTION
**Pull request by jviau**
Adds on optional JsonConverter to support forward compatibility with `OrchestrationInstance` in AzureStorage provider. This is enabled via a setting. I opted for the setting to minimize any impact on anyone upgrading to this version.

Closes #491

### CODE REVIEW

File: src/DurableTask.AzureStorage/OrchestrationInstanceConverter.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            if (objectType == null)
+            {
+                throw new ArgumentNullException(nameof(objectType));
+            }
+
+            // Variant #2 incase this needs to support any extensible data contract.
+            //return objectType.GetCustomAttribute<DataContractAttribute>() != null
+            //    && typeof(IExtensibleDataObject).IsAssignableFrom(objectType);
+
+            return typeof(OrchestrationInstance).IsAssignableFrom(objectType);
```
Comment (cgillum) [inline src/DurableTask.AzureStorage/OrchestrationInstanceConverter.cs:34]: As far as I can tell, nothing in the serialization or deserialization logic depends on the object being an instance of `OrchestrationInstance`. It seems to me you could use this for any data type. Can we rename this class to `DataContractJsonConverter` and remove this type constraint? That would also make the serializer better match the name of the property that enables it (`UseDataContractSerialization`).

**Review by cgillum — APPROVED** [review]
LGTM!

**Comment by jviau** [comment]
Making the change to convert more than just the `OrchestrationInstance` is causing some problems. I'll look into either fixing, or scoping it back down to just that type.

**Comment by jviau** [comment]
Resolved the issues:
1. DateTime format was causing issues. This now uses the DateTimeFormat from the json serializer. This is not perfect though - if they have a custom converter for date time, it may cause issues. But then again, this should _replace_ netwonsoft json serialization for these types and not interop with it.
2. `GenericEvent` was missing from `HistoryEvent.KnownTypes()` - refactored that method to use reflection to load up the initial list of types (then it is cached).

**Review by cgillum — APPROVED** [review]
Latest changes look great! Thanks for doing the extra validation and making the necessary fixes.

**Pull request merged**